### PR TITLE
Fix MacOS test flakes by releasing cached MPS memory between tests

### PR DESCRIPTION
**Pull request by anuragg1209**
## Issue
Please link the corresponding GitHub issue. If an issue does not already exist,
please open one to describe the bug or feature request before creating a pull request.

This allows us to discuss the proposal and helps avoid unnecessary work.

## Motivation and Context

PyTorch's MPS caching allocator holds freed memory for the process lifetime. On ~7GB macos-latest runners the cache accumulates across tests (finetuning tests move ~1GB of model weights, optimizer state, and eval clones through MPS per parametrization) until the ~3.3 GiB limit is hit and unrelated tests OOM on KiB-sized allocations. Add an autouse fixture running gc.collect() + torch.mps.empty_cache() after each test; the gc pass is needed because autospec'd mocks keep MPS tensors alive in reference cycles. Fixes PRI-339.

## Public API Changes

-   [x] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

Tested locally on an MPS device, and the MPS cache is now released between tests. 

## Checklist

-   [ ] The changes have been tested locally.
-   [ ] Documentation has been updated (if the public API or usage changes).
-   [ ] A changelog entry has been added (see `changelog/README.md`), or "no changelog needed" label requested.
-   [ ] The code follows the project's style guidelines.
-   [ ] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -17,3 +19,19 @@ def set_global_seed() -> None:
...
+    PyTorch's MPS caching allocator holds freed memory for the lifetime of the
+    process. On the ~7GB macos-latest CI runners the cache accumulates across
+    tests until the ~3.3 GiB MPS limit is hit and unrelated tests OOM.
+    gc.collect() first so tensors kept alive by reference cycles (e.g.
+    unittest.mock call records) are actually freed before emptying the cache.
+    """
+    yield
+    if torch.backends.mps.is_available():
+        gc.collect()
+        torch.mps.empty_cache()
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply tests/conftest.py:37]: This should not be an issue, as in the dependencies we enforce `"torch>=2.5"`. Closing this issue, but feel free to open it if this needs to be addressed.

File: tests/conftest.py [hunk 2]
```diff
@@ -17,3 +19,19 @@ def set_global_seed() -> None:
     torch.manual_seed(seed)
     np.random.seed(seed)  # noqa: NPY002
     random.seed(seed)
+
+
+@pytest.fixture(autouse=True)
+def release_mps_memory() -> Generator[None]:
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply tests/conftest.py:25]: Generators' send/return params default to `None` per PEP 696, the repo's pinned `mypy 2.1.0` passes on this file. Closing this issue.

**Review by oscarkey — APPROVED** [review]
amazing, thank you!
